### PR TITLE
Recommender systems

### PR DESCRIPTION
**Pull request by gathurian**

### CODE REVIEW

File: FS19/ML/Zusammenfassung_ML.tex [hunk 1]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+        6(3x + 2y)\\
+        4(3x + 2y)
+    \end{bmatrix}\\
+        \nabla f(1,2) &= \begin{bmatrix}
+            42\\
+            28
+        \end{bmatrix}
+        \end{align*}
+
+        \subsection{Properties of the gradient}
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: Your autoformatter is drunk, send him home, but seriously, the formatting right now makes it hard to read, I agree on indenting equations and figures, but sections / subsection, hell naw dawg...

File: FS19/ML/Zusammenfassung_ML.tex [hunk 2]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+                \end{itemize}
+            \item Pick the model $M-K$ with the lowest estimated generalisation error and retrain that model on the entirety of the training set $S$.
+        \end{enumerate}
+
+        \section{Recommender Systems}
+
+        \subsection{Definition}
+
+        \begin{quote}
+            \blockquote[Burke, 2002]{A recommender system is any system that produces indiviudalized recommendations as output or has the effect of guiding the user in a personalized way to interesting or useful objects in a large space of possible options}
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: individualized*

File: FS19/ML/Zusammenfassung_ML.tex [hunk 3]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+        \end{quote}
+
+        Or easier said:
+
+        \begin{quote}
+            \blockquote[Jennach et. al., 2010]{A recommender system is a software application capable of suggesting interesing things to its users after learning their preferences over time}
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: interesting* or [sic]

File: FS19/ML/Zusammenfassung_ML.tex [hunk 4]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+        \end{quote}
+
+        So basically recommender systems are the algorithms that show up under your amazon searches as 'customers who bought [xxx] also bought [yyy]'. 
+
+        As opposed to Support or Lift, recommender systems are \textbf{indiviudalized}.
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: individualised

File: FS19/ML/Zusammenfassung_ML.tex [hunk 5]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+
+        \vspace{10px}
+
+        The goal of recommender systems is to give customers a better overview, to help them cope with the information overload problem (Amazon currently has over 20'000 results listed of 'Laptop Stand', but I only need one).
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: results ~listed~ of 'Laptop Stand'

File: FS19/ML/Zusammenfassung_ML.tex [hunk 6]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+            \includegraphics[keepaspectratio=true,height=15\baselineskip]{recommender_system.png}
+            \caption{Generic Picture of a Commercial Recommender System}
+            \label{fig:recommender_systems}
+        \end{figure}
+
+
+        \subsection{Business Model}
+
+        \begin{itemize}
+            \item Recommender Systems are only proviced as SaaS (Software as a Service), you only rent the system
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: only provided*

File: FS19/ML/Zusammenfassung_ML.tex [hunk 7]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
-	\item Make sure the features are on a similar scale and get every feature into the range of $-1\leq x_i \leq 1$
-	\begin{enumerate}
-		\item Replace $x_i$ with $x_i - \bar{x}_i$ to have features with an approximately zero mean. Does not apply to $x_0 = 1$
-		\item Divide $x_i - \bar{x}_i$ by the range of $x_i = max(x_i) - min(x_i)$ or the standard deviation of $x_i$ 
-	\end{enumerate}
-	\item Test if Gradient Descent is working correctly by plotting the cost function $J(\bm{\theta})$ versus the number of iterations
-	\item Declare convergence if $J(\bm{\theta})$ decreases by less than a given threshold (for example $10^-3$) in one iteration
-\end{enumerate}
+        \begin{equation}
+            \dfrac{\dfrac{X\ and\ Y}{X}}{\dfrac{¬X\ and\ Y}{¬X}}
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: 1. Use \frac not \dfrac - https://tex.stackexchange.com/a/135395
2. ¬ is not recognised as a valid character use \neg

File: FS19/ML/Zusammenfassung_ML.tex [hunk 8]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
-\end{equation}
+        To summarize:
 
-The second sum goes from $1$ to the number of features $m$ and it does not contain $\theta_0$. The regularisation hyperparameter $\lambda$ controls the amount of regularisation. If $\lambda = 0$ there will be no regularization, and if $\lambda = \infty$ there will be underfitting because only $\theta_0$ will be different from zero.
-In Ridge regularised regression, we choose $\theta$ to minimize $J(\theta)$:
-\begin{equation}
-	\hat{\theta} = \underset{\theta}{\text{argmin}} J(\theta)
-\end{equation}
+        \begin{itemize}
+            \item Recommendations by Associations are \textit{fast}. No data collection and evaluation necessa4ry
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: necessary*

File: FS19/ML/Zusammenfassung_ML.tex [hunk 9]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+            \label{fig:context_based_recommendations}
+        \end{figure}
 
-\subsubsection{k-fold Cross Validation}
+        \begin{figure}[htb!]
+            \centering
+            \includegraphics[keepaspectratio=true, width=0.9\textwidth]{item_similarity.png}
+            \caption{Item Similarity}
+            \label{fig:item_similarity}
+        \end{figure}
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: Would it be possible to have a white background in these pictures?
grafik

File: FS19/ML/Zusammenfassung_ML.tex [hunk 10]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+            \item Collect as much user preferences as you can (e.g. by ratings, wish lists, item viewed etc.)
+            \item Compute predictions for products labelled with $?$ based on item similarity (See Equation \ref{eq:sim} on page \pageref{eq:sim})
+            \item Recommend products with high prediction value
+        \end{enumerate}
+
+        \newpage
+
+        \begin{wrapfigure}[14]{R}{0.5\textwidth}
+            \centering
+            \includegraphics[keepaspectratio=true,height=10\baselineskip]{context_based_recommendations_v2.png}
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: grafik
Not readable

File: FS19/ML/Zusammenfassung_ML.tex [hunk 11]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+        \begin{figure}[htb!]
+            \centering
+            \includegraphics[keepaspectratio=true, width=0.9\textwidth]{context_based_recommendations_v3.png}
+            \caption{Similarity between books}
+            \label{fig:book_similarity}
+        \end{figure}
+
+        \begin{figure}[htb!]
+            \centering
+            \includegraphics[keepaspectratio=true, width=0.9\textwidth]{context_based_recommendations_v4.png}
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: grafik
Background?

File: FS19/ML/Zusammenfassung_ML.tex [hunk 12]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+        \vspace{10px}
+
+        Neighbors $N_j$ refer to the (at most) $N$ rated items most similar to item $j$, $N$ being a hyperparameter that can be tweaked for improved algorithm-performance
+        \noindent Select only rated items with similarity $> 0$ (choose less than $N$ if you cannot find enough candidates)
+
+        \vspace{50px}
+
+        \begin{figure}[htb!]
+            \centering
+            \includegraphics[keepaspectratio=true, width=0.9\textwidth]{context_based_recommendations_v3.png}
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: grafik
Higher DPI please

File: FS19/ML/Zusammenfassung_ML.tex [hunk 13]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+
+        \begin{figure}[htb!]
+            \centering
+            \includegraphics[keepaspectratio=true, width=\linewidth]{item_to_item_collaborative_filtering.png}
+            \caption{Item to Item Collaborative Filtering}
+            \label{fig:item_to_item_collaborative_filtering}
+        \end{figure}
+
+
+        Advantages of Collaborative Fiiltering Recommendations
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: Filtering*

File: FS19/ML/Zusammenfassung_ML.tex [hunk 14]
```diff
@@ -1256,236 +1256,558 @@ \section{Gradient Descent}
...
+
+        To improve the recommendations for sparsely filled matrices, there's something called \textbf{Low Rank Matrix Factorization}. The matrix we saw before (Fig. \ref{fig:item_to_item_collaborative_filtering}, 
+        \ref{fig:user_similarity}, \ref{fig:book_similarity}) can be split up in two matrices, which are more densely filled. 
+
+        One matrix holds all users and their ratings for the products they rated and the second matrix holds all products.
+
+
+        The first matrix (U) holds all users and their ratings. Each row is a user and all the ratings on all products they have made.
+
+        The second matrix (V) holds all the products and the rating they've gotten. This matrix is \textbf{transposed}. o
```
Comment (KilnOfTheSecondFlame) [inline FS19/ML/Zusammenfassung_ML.tex]: This matrix is \textbf{transposed}. ~o~

**Review by KilnOfTheSecondFlame — CHANGES_REQUESTED** [review]
Use higher DPI images and look that your desktop image does not show up in the background of said images. Also please not so much indentation D:

File: FS19/ML/Zusammenfassung_ML.tex (hunk 1 above)
Comment (gathurian) [reply FS19/ML/Zusammenfassung_ML.tex]: Fixed by switching to an inferior, semi-proprietary editor.... Look at what you made me do

File: FS19/ML/Zusammenfassung_ML.tex (hunk 10 above)
Comment (gathurian) [reply FS19/ML/Zusammenfassung_ML.tex]: Replaced with hi-res img

File: FS19/ML/Zusammenfassung_ML.tex (hunk 11 above)
Comment (gathurian) [reply FS19/ML/Zusammenfassung_ML.tex]: Replaced with hi-res img without transparancy (but my background is dope)

File: FS19/ML/Zusammenfassung_ML.tex (hunk 12 above)
Comment (gathurian) [reply FS19/ML/Zusammenfassung_ML.tex]: Replaced with hi-res img

**Review by KilnOfTheSecondFlame — APPROVED** [review]

**Pull request merged**